File: fastapi/app/cases/case1_lock/lock_service.py
import asyncio
import uuid
from redis.asyncio import Redis

# Lua script to safely release lock only if the token matches (prevents releasing expired locks held by others)
RELEASE_LOCK_LUA = """
if redis.call("get", KEYS[1]) == ARGV[1] then
    return redis.call("del", KEYS[1])
else
    return 0
end
"""


class DistributedLock:
    def __init__(self, redis: Redis, resource: str, ttl_ms: int = 5000, retry_delay_ms: int = 50, max_retries: int = 20):
        self.redis = redis
        self.resource = resource
        self.lock_key = f"lock:{resource}"
        self.ttl_ms = ttl_ms
        self.retry_delay_ms = retry_delay_ms
        self.max_retries = max_retries
        self.token: str | None = None

    async def acquire(self) -> bool:
        token = str(uuid.uuid4())
        for _ in range(self.max_retries):
            # SET lock:resource token NX PX ttl_ms
            acquired = await self.redis.set(self.lock_key, token, px=self.ttl_ms, nx=True)
            if acquired:
                self.token = token
                return True
            await asyncio.sleep(self.retry_delay_ms / 1000.0)
        return False

    async def release(self) -> bool:
        if not self.token:
            return False
        result = await self.redis.eval(RELEASE_LOCK_LUA, 1, self.lock_key, self.token)
        return bool(result)

    async def __aenter__(self):
        acquired = await self.acquire()
        if not acquired:
            raise TimeoutError(f"Could not acquire lock for resource '{self.resource}'")
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.release()


async def init_inventory(redis: Redis, item_id: str, stock: int) -> None:
    await redis.set(f"stock:{item_id}", stock)


async def get_inventory(redis: Redis, item_id: str) -> int:
    val = await redis.get(f"stock:{item_id}")
    return int(val) if val is not None else 0


async def purchase_with_lock(redis: Redis, item_id: str, quantity: int = 1) -> dict:
    async with DistributedLock(redis, f"inventory:{item_id}"):
        current_stock = await get_inventory(redis, item_id)
        if current_stock < quantity:
            return {"success": False, "message": "Out of stock", "remaining": current_stock}
        # Artificial slight delay to simulate business validation / DB latency
        await asyncio.sleep(0.01)
        new_stock = current_stock - quantity
        await redis.set(f"stock:{item_id}", new_stock)
        return {"success": True, "message": "Purchase successful", "remaining": new_stock}


async def purchase_without_lock(redis: Redis, item_id: str, quantity: int = 1) -> dict:
    # Intentionally vulnerable to Race Condition for learning demonstration
    current_stock = await get_inventory(redis, item_id)
    if current_stock < quantity:
        return {"success": False, "message": "Out of stock", "remaining": current_stock}
    # During this delay, concurrent requests read the same stale current_stock!
    await asyncio.sleep(0.01)
    new_stock = current_stock - quantity
    await redis.set(f"stock:{item_id}", new_stock)
    return {"success": True, "message": "Purchase successful", "remaining": new_stock}

File: nestjs/src/cases/case1-lock/lock.service.ts
import { Injectable } from '@nestjs/common';
import { randomUUID } from 'crypto';
import { RedisService } from '../../core/redis/redis.service';

// 락을 안전하게 해제하기 위한 Lua 스크립트.
// 현재 락 키의 값이 획득 시 발급된 토큰과 일치할 때만 원자적으로 삭제하여
// 타 요청에 의해 새로 획득된 락을 오삭제하지 않도록 보장합니다.
const RELEASE_LOCK_LUA = `
if redis.call("get", KEYS[1]) == ARGV[1] then
    return redis.call("del", KEYS[1])
else
    return 0
end
`;

/**
 * Redis 기반 분산 락 및 재고 관리 서비스.
 * 다중 인스턴스 환경에서 공유 자원(상품 재고)에 대한 동시성 제어를 담당합니다.
 */
@Injectable()
export class LockService {
  constructor(private readonly redisService: RedisService) {}

  private get redis() {
    return this.redisService.getClient();
  }

  /**
   * Redis SET NX PX를 활용한 분산 락 획득 시도.
   *
   * @param resource 락 대상 자원 식별자
   * @param ttlMs 락 만료 시간 (기본값: 5000ms)
   * @param maxRetries 최대 재시도 횟수 (기본값: 20회)
   * @param retryDelayMs 재시도 대기 간격 (기본값: 50ms)
   * @returns 락 획득 성공 시 고유 토큰 문자열, 실패 시 null
   */
  async acquireLock(resource: string, ttlMs = 5000, maxRetries = 20, retryDelayMs = 50): Promise<string | null> {
    const token = randomUUID();
    const lockKey = `lock:${resource}`;

    for (let i = 0; i < maxRetries; i++) {
      // SET lock:resource token PX ttlMs NX (원자적 락 선점)
      const result = await this.redis.set(lockKey, token, 'PX', ttlMs, 'NX');
      if (result === 'OK') {
        return token;
      }
      await new Promise((r) => setTimeout(r, retryDelayMs));
    }
    return null;
  }

  /**
   * 보유한 고유 토큰을 검증하여 분산 락을 안전하게 해제.
   */
  async releaseLock(resource: string, token: string): Promise<boolean> {
    const lockKey = `lock:${resource}`;
    const result = await this.redis.eval(RELEASE_LOCK_LUA, 1, lockKey, token);
    return result === 1;
  }

  /** 상품 재고 수량 초기화 */
  async initInventory(itemId: string, stock: number): Promise<void> {
    await this.redis.set(`stock:${itemId}`, stock.toString());
  }

  /** 현재 상품 재고 수량 조회 */
  async getInventory(itemId: string): Promise<number> {
    const val = await this.redis.get(`stock:${itemId}`);
    return val !== null ? parseInt(val, 10) : 0;
  }

  /**
   * [동시성 안전] 분산 락을 적용한 재고 차감 구매 로직.
   * 락 획득 -> 재고 조회 -> 검증 -> 차감 -> 락 해제(finally) 흐름을 보장합니다.
   */
  async purchaseWithLock(itemId: string, quantity = 1): Promise<{ success: boolean; message: string; remaining: number }> {
    const token = await this.acquireLock(`inventory:${itemId}`);
    if (!token) {
      throw new Error('Lock acquisition timeout');
    }

    try {
      const currentStock = await this.getInventory(itemId);
      if (currentStock < quantity) {
        return { success: false, message: 'Out of stock', remaining: currentStock };
      }
      // 인위적 DB 처리 지연 시뮬레이션
      await new Promise((r) => setTimeout(r, 10));
      const newStock = currentStock - quantity;
      await this.redis.set(`stock:${itemId}`, newStock.toString());
      return { success: true, message: 'Purchase successful', remaining: newStock };
    } finally {
      await this.releaseLock(`inventory:${itemId}`, token);
    }
  }

  /**
   * [동시성 취약] 분산 락 미적용 구매 로직 (Race Condition 학습 및 시연용).
   * 동시 다발적 요청 인입 시 동일한 현재 재고를 읽어 초과 판매가 발생합니다.
   */
  async purchaseWithoutLock(itemId: string, quantity = 1): Promise<{ success: boolean; message: string; remaining: number }> {
    const currentStock = await this.getInventory(itemId);
    if (currentStock < quantity) {
      return { success: false, message: 'Out of stock', remaining: currentStock };
    }
    // 동시성 결함 발생 창(Race condition window)
    await new Promise((r) => setTimeout(r, 10));
    const newStock = currentStock - quantity;
    await this.redis.set(`stock:${itemId}`, newStock.toString());
    return { success: true, message: 'Purchase successful', remaining: newStock };
  }
}
